docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
+- Support for Helm install, upgrade, test, rollback and uninstall
+- Performs Helm uninstall on `HelmRelease` CR removal
```

Comment:
**From my perspective, the fact that Flux uninstalls the HelmChart when a flux HelmRelease is deleted is expected but not an advantage in our case and I think it should be considered a drawback.** This scenario exemplifies the limitations and complexities that may arise when adopting Flux, potentially restricting our control over our solutions according to our specific needs. It restricts our ability to control the deletion process and ensure that essential operations are completed before uninstalling the solution.

IMO: The fundamental purpose of our Operator is to establish a control-loop that triggers whenever a change occurs. Its primary objective is to ensure, through a consistent and idempotent set of operations, that the desired state matches the current state. This control-loop serves as a mechanism for continuously monitoring and maintaining the system in line with our defined configurations and expectations. 

For instance, if a user mistakenly attempts to delete an AddOn CR, we invoke a  [finalizer](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) function that ensures the AddOn CR is only deleted once all necessary operations, such as uninstalling the HelmChart, have been successfully completed. You can refer to this example in our prototype [here](https://github.com/replicatedhq/brainstorming-ekco/pull/2/files#diff-4d05894b38940f59fcac351cf0cfb3f1ae42eddc423299655677f122dbe4bfe2R299-R343).
4d05894b38940f59fcac351cf0cfb3f1ae42eddc423299655677f122dbe4bfe2R299-R343) an example our prototype.  

However, when using Flux, if a user accidentally deletes the HelmRelease CR that Flux uses for managing Helm releases, it will uninstall the HelmChart solution through the FluxHelmController. In such a scenario, our control-loop can only recreate the Flux HelmRelease but cannot prevent this situation or guarantee that required backups or migrations, for example, have occurred, as we could if we relied solely on Helm. Unless would we be able to say to Flux that it can only uninstall the HelmRelease when a flag that it can be done by adding some info into its HelmRelease CR to prevent this scenario. Is that possible with Flux?

Reply:
Can you provide an example when it's beneficial to not delete the CR? I can't think of a situation and it would be helpful to understand why you see it beneficial to write custom deletion code ourself.

Reply:
@chris-sanders 

The question isn't about choosing not to delete the Custom Resource (CR), but about ensuring that all necessary precautions for a safe uninstallation are undertaken before deleting a CR, which represents an Add-On and its corresponding solution in the cluster.

For example, consider a scenario where the OpenEBS Kurl/Installer CR is deleted manually or via any other solution, whether by mistake or intentionally. Before proceeding with the Helm uninstallation of OpenEBS, we could guarantee that an alternative storage solution is already in place and that the volume data has been successfully migrated to this new solution. Otherwise, the OpenEBS CR will not get deleted unless you explicitly force it manually. Furthermore, we could for example ensure that backup solution should be employed to store a copy of the data. This will allow data recovery via support, if necessary.

In essence, the focus is not on whether or not to delete the CR but rather on ensuring that all the required safety checks and procedures are completed prior to its deletion. 

Same is valid if someone try to delete the namespace where the AddON CR leaves. By using finalizers the namespace cannot get deleted prior all resources on it be first get deleted. To better understand you can check the how finalizers work in the k8s docs: https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/ and see how the k8s api and us building an Operator would leverage in the cascading delation: https://kubernetes.io/docs/tasks/administer-cluster/use-cascading-deletion/